`CipherListView` updates

nick-livefront opened the pull request:
## 🎟️ Tracking

Tangential to [PM-22134](https://bitwarden.atlassian.net/browse/PM-22134)

## 📔 Objective

Two updates to ensure parity when migrating from `CipherView` to `CipherListView` on the clients repo
- Added `local_data` so the type propagates to the clients repo and can be set there.
- Added  `CopiableCipherFields` enum + property.
  - In the browser the copy menu / actions are dynamic based on the populated fields on the cipher.  The `CipherListView` doesn't contain these details. This array is filled with the available fields so the extension can still check their availability. When a value is needed the full cipher is decrypted. 
- Added `has_old_attachments` to identify attachments following old encryption methods. This populates UI signaling to the user that the attachment needs to be re-encrypted. 

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes


[PM-22134]: https://bitwarden.atlassian.net/browse/PM-22134?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nick-livefront committed:
initial add of copiable fields

nick-livefront committed:
update copiable fields to be an enum array

nick-livefront committed:
add local_data to CipherListView

nick-livefront committed:
update comment

nick-livefront committed:
fix tests

Hinton commented on crates/bitwarden-vault/src/cipher/cipher.rs:
```diff
@@ -209,6 +227,14 @@ pub struct CipherListView {
+    pub copiable_fields: Vec<CopiableCipherFields>,
+
+    // Only added to the list view for typing purposes, it won't be retrieved from the server
+    pub local_data: Option<LocalDataView>,
```
**question:** Is this be able to access `last_used_date` and `last_launched`? I think we should wire it up similar to how it's done on `CipherView`. Currently `local_data` would be empty since it's always empty on `Cipher` but if it starts being populated it will have the same behaviour for `CipherListView` and `CipherView`.

nick-livefront replied:
Good catch, yes! Those details are needed for sorting in the lists. I updated the logic to match `CipherView`. [eb5a944](https://github.com/bitwarden/sdk-internal/pull/306/commits/eb5a9448041efcd33f5b0bfd993bde42228e18d3)

Hinton commented on crates/bitwarden-vault/src/cipher/cipher.rs:
```diff
@@ -176,6 +176,24 @@ pub enum CipherListViewType {
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
+pub enum CopiableCipherFields {
```
**nit**: I think `Copyable` is more commonly used in modern english. If we keep `Copiable` we should add it to the wordlist, `.vscode/settings.json` since it's currently not detected as a valid word.

nick-livefront replied:
Agreed: [b2694b6](https://github.com/bitwarden/sdk-internal/pull/306/commits/b2694b68e8790c22fd7ceadc61706fab2adf3396)

Hinton commented on crates/bitwarden-vault/src/cipher/cipher.rs:
```diff
@@ -409,6 +435,102 @@ impl Cipher {
+
+    /// Returns a list of copiable field names for this cipher,
+    /// based on the cipher type and populated properties.
+    fn get_copiable_fields(&self) -> Vec<CopiableCipherFields> {
```
**issue:** This file actually contains some anti patterns related to subtitle, and this feature is a bit rough to test right now.

**suggestion:** This PR motivated me to revisit subtitle and I created https://github.com/bitwarden/sdk-internal/pull/307. We should lift up the implementations into the cipher type structs in a similar way.

**suggestion:** I think you can get rid of the mutability, especially when extracted to smaller functions you can have something like:

```rust
[
    card.number.map(|_| CopiableCipherFields::CardNumber),
    card.code.map(|_| CopiableCipherFields::CardSecurityCode),
]
.into_iter()
.filter_map(|opt| opt)
.collect()
```

nick-livefront replied:
Oh nice! This is way better, I really like the delegation into the sub types. [fe0e02c](https://github.com/bitwarden/sdk-internal/pull/306/commits/fe0e02caead9ceab92151b0e21be9de66a7c2295) I will pull your changes in and refactor to use `getKind` after your PR merges.

On the rust side of things I would consider myself still learning so if there are other pragmatic fixes that I can make here, feel free to point them out! I'm happy to learn on the fly here.

nick-livefront committed:
decrypt local_data for the `CipherListView`

nick-livefront committed:
"copiable" to "copyable"

nick-livefront committed:
refactor `getCopyableFields` into each type of cipher for a nicer separation of concerns

nick-livefront committed:
formatting

nick-livefront committed:
add `has_old_attachments` boolean to `CipherListView`

nick-livefront commented:
@Hinton I rebased off of `main` to utilize your changes and I added another attribute I missed: `has_old_attachments`. 

This is good for another review!

Hinton commented on crates/bitwarden-vault/src/cipher/card.rs:
```diff
@@ -233,4 +291,40 @@ mod tests {
+            number: None,
+        };
+
+        let cipher = create_cipher_for_card(card.clone());
```
**suggestion:** I think we can just skip creating a Cipher and call this directly on `card`? Avoids needing all the boilerplate code for creating ciphers.

nick-livefront replied:
For all cipher types except `SecureNotes` this is true. The `SecureNotes` type references the `notes` property on the cipher. But you're right, this is just added boilerplate for all other cipher types. I changed the parameter to an `Option` and then removed the boilerplate from the tests that do not utilize it. [a8035b8](https://github.com/bitwarden/sdk-internal/pull/306/commits/a8035b8ee7aa35041953fa174c3d7fbbee898834)

Hinton replied:
Ah, notes 😠.

Hinton commented on crates/bitwarden-vault/src/cipher/card.rs:
```diff
@@ -233,4 +291,40 @@ mod tests {
+            cardholder_name: None,
+            exp_month: None,
+            exp_year: None,
+            code: Some(encrypt_test_string("123")),
```
**suggestion:** We never decrypt these so we can just use any valid enc string? Saves some more boilerplate in tests.

nick-livefront replied:
Agreed [8cea293](https://github.com/bitwarden/sdk-internal/pull/306/commits/8cea29354f9dbe3e3499bc22fc600a8771181804)

Is there an easy way to generate encrypted strings? I was using `println` to retrieve an actual encrypted string, maybe there is a better way?

Hinton replied:
You could copy them from the web apps or cli.

Hinton commented on crates/bitwarden-vault/src/cipher/cipher.rs:
```diff
@@ -219,6 +240,16 @@ pub struct CipherListView {
+
+    /// Fields on the cipher that are populated. This can be used in the
+    /// UI to determine the visibility of copy actions without needing
+    /// the full cipher details.
```
**suggestion (nit):** We can be a bit more succinct without losing context.

```suggestion
    /// Hints for the presentation layer for which fields can be copied.
```

nick-livefront replied:
[ac7f600](https://github.com/bitwarden/sdk-internal/pull/306/commits/ac7f6004f8e67fada62fa14f2986a0505d9b4ec1)

Hinton commented on crates/bitwarden-vault/src/cipher/cipher.rs:
```diff
@@ -219,6 +240,16 @@ pub struct CipherListView {
+    pub copyable_fields: Vec<CopyableCipherFields>,
+
+    /// Indicates if the cipher has old attachments that need to be re-uploaded
+    pub has_old_attachments: bool,
```
**suggestion (nit):** Move this up to `attachments`

nick-livefront replied:
I took this as, move it up next to the attachments property. Let me know if I misread that! [12bf038](https://github.com/bitwarden/sdk-internal/pull/306/commits/12bf038885635767bd11eb97c3275576e38eada7)

Hinton commented on crates/bitwarden-vault/src/cipher/cipher.rs:
```diff
@@ -380,6 +411,38 @@ impl Cipher {
+                .as_ref()
+                .map(|secure_note| secure_note.get_copyable_fields(self))
+                .unwrap_or_default(),
+        }
```
**suggestion:** You should be able to use `get_kind` here and simplify it. `get_subtitle` should be a good example:

```suggestion
        self.get_kind()
            .map(|sub| sub.get_copyable_fields())
            .unwrap_or_default()
```

nick-livefront replied:
🤦 [6d424ea](https://github.com/bitwarden/sdk-internal/pull/306/commits/6d424ea49732fdf4d0caba8e454c1d164bd9bce2)

nick-livefront committed:
use `getKind` rather than pattern matching

nick-livefront committed:
all cipher to be optionally passed in to `get_copyable_fields`

nick-livefront committed:
update comment to be more succinct

nick-livefront committed:
move `has_old_attachments` to be grouped with `attachments` property

nick-livefront committed:
use hard coded encrypted strings because the underlying decrypted value isn't needed

Hinton reviewed (approved):
Thanks for iterating on this. Looks good to me.

gbubemismith reviewed (approved):
This definitely makes things easier on the client 🚀
